### DISCUSSION TITLE
auxvariable and non linear variable

### DISCUSSION THREAD

**Discussion by mcacace** [idaholab/moose#16368] [Q&A General] [answered]
Dear all,
I would have a simple (naive) questions: is it possible to add to a non linear variable the value computed from an aux variable? And, if yes, when and where would it be done? One thing to add is that what I would need is actually the coupleGradient of the non linear variable afterwards.
Thanks for any help,
mauro

**Comment by WilkAndy** [comment]
Yes, that is possible, but it'd be useful to know more details:

- are you constructing a new AuxVariable = anotherAux + u ?
- are you doing this in the input file, or in C++ code?
- instead, are you wanting to modify u, eg, u = u + Aux?   If so, when do you want the modification done?  After every nonlinear iteration, after every timestep, ... ?   Are you doing this in C++ code, or using the input file (I believe it is impossible to do this from the input file)?

**Reply by mcacace** [reply]
Dear @WilkAndy,
Thanks. Mine would exactly match your last bullet point. I compute a non linear var (u) and I have an aux kernel that provides an aux and would like to have it as a correction to u (u=u+aux) at timestep end (do not need to internally iterate). Indeed, I would only want to make use of the new value (its coupledgradient actually) in the next timestep. All would be in a c++ file on runtime, though I still did not find exactly where and how. As always, any suggestion is greatly appreciated. 
Thanks again, 
Mauro

**Reply by WilkAndy** [reply]
I suggest you attempt to write your DEs so the AuxVariable is included in the DEs, and hence just becomes a coupledVariable in a Kernel.   That will fit nicely into the MOOSE paradigm, and could be more physical too.  Eg, maybe the AuxVariable is some sort of source term.

If that just isn't possible, then you could follow the code at

https://github.com/idaholab/moose/blob/next/modules/richards/src/base/RichardsMultiphaseProblem.C

which modifies a nonlinear variable in a similar way to what you'd like.

**Reply by mcacace** [reply]
@WilkAndy: I had a quick look at the code (above) in the Richards' module. One question: should it only work based on two non linear variables (and not with one non linear and one aux variable)? Just wondering, sorry for posting back...

**Reply by WilkAndy** [reply]
Yea, it'll probably work fine.

**Comment by dschwen** [comment]
[CoupledForce](https://mooseframework.inl.gov/source/kernels/CoupledForce.html) might be what you're looking for.

**Comment by mcacace** [comment, marked as answer]
Thanks @WilkAndy and @WilkAndy. I ended up with something similar on CoupleForce. Will nevertheless try to have a deeper look at the original code in the framework. 
mauro
